airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,170 @@
+
+    if (checkpointOffsetToSend.isEmpty()) {
+      try {
+        final HashMap<String, String> temporalOffset = (HashMap<String, String>) offsetManager.read();
+        if (!targetPosition.isSameOffset(previousCheckpointOffset, temporalOffset)) {
+          checkpointOffsetToSend.putAll(temporalOffset);
+        }
+      } catch (final ConnectException e) {
+        LOGGER.warn("Offset file is being written by Debezium. Skipping CDC checkpoint in this loop.");
+      }
+    }
+
+    if (checkpointOffsetToSend.size() == 1
+        && !message.isSnapshotEvent()) {
+      if (targetPosition.isEventAheadOffset(checkpointOffsetToSend, message)) {
+        sendCheckpointMessage = true;
+      } else {
+        LOGGER.info("Encountered records with the same event offset.");
+      }
+    }
+
+    return eventConverter.toAirbyteMessage(message);
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(ConfiguredAirbyteStream stream) {
+    final var syncFinishedOffset = (HashMap<String, String>) offsetManager.read();
+
+    System.out.println("initialoffset: " + initialOffset);
+    System.out.println("syncFinishedOffset: " + syncFinishedOffset);
+
+    if (targetPosition.isSameOffset(initialOffset, syncFinishedOffset)) {
+      // Edge case where no progress has been made: wrap up the
+      // sync by returning the initial offset instead of the
+      // current offset. We do this because we found that
+      // for some databases, heartbeats will cause Debezium to
+      // overwrite the offset file with a state which doesn't
+      // include all necessary data such as snapshot completion.
+      // This is the case for MS SQL Server, at least.
+      return createStateMessage(initialOffset);
+    }
+    return createStateMessage(syncFinishedOffset);
+  }
+
+  private void resetCheckpointValues() {
+    checkpointOffsetToSend.clear();
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public boolean shouldEmitStateMessage(ConfiguredAirbyteStream stream) {
+    return cdcStateHandler.isCdcCheckpointEnabled() && sendCheckpointMessage;
+  }
+
+  /**
+   * Creates {@link AirbyteStateMessage} while updating CDC data, used to checkpoint the state of the
+   * process.
+   *
+   * @return {@link AirbyteStateMessage} which includes offset and schema history if used.
+   */
+  private AirbyteStateMessage createStateMessage(final Map<String, String> offset) {
+    if (trackSchemaHistory && schemaHistoryManager == null) {
+      throw new RuntimeException("Schema History Tracking is true but manager is not initialised");
+    }
+    if (offsetManager == null) {
+      throw new RuntimeException("Offset can not be null");
+    }
+
+    final AirbyteStateMessage message =
+        cdcStateHandler.saveState(offset, schemaHistoryManager != null ? schemaHistoryManager.read() : null).getState();
```

Comment:
Wrap the `schemaHistoryManager` type in an `Optional<>` to avoid this kind of ugliness.